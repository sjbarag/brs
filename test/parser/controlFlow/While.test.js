const brs = require("brs");
const { Lexeme } = brs.lexer;
const { BrsBoolean, BrsString, Int32 } = brs.types;
const { token, identifier, EOF } = require("../ParserTests");

describe("parser while statements", () => {
    let parser;

    beforeEach(() => {
        parser = new brs.parser.Parser();
    });

    test("while without exit", () => {
        const { statements, errors } = parser.parse([
            token(Lexeme.While, "while"),
            token(Lexeme.True, "true", BrsBoolean.True),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.Print, "print"),
            token(Lexeme.String, "looping", new BrsString("looping")),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.EndWhile, "end while"),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    test("while with exit", () => {
        const { statements, errors } = parser.parse([
            token(Lexeme.While, "while"),
            token(Lexeme.True, "true", BrsBoolean.True),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.Print, "print"),
            token(Lexeme.String, "looping", new BrsString("looping")),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.ExitWhile, "exit while"),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.EndWhile, "end while"),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    test("nested", () => {
        const { tokens } = brs.lexer.Lexer.scan(
            `
            while i < 1000
                while j < 1000
                    while k < 1000 : k++ : end while
                    j++
                end while
                i++
            end while
            `
        );

        const { statements, errors } = parser.parse(tokens);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    test("location tracking", () => {
        /**
         *    0   0   0   1   1
         *    0   4   8   2   6
         *  +------------------
         * 1| while true
         * 2|   Rnd(0)
         * 3| end while
         */
        const { statements, errors } = parser.parse([
            {
                kind: Lexeme.While,
                text: "while",
                isReserved: true,
                location: {
                    start: { line: 1, column: 0 },
                    end: { line: 1, column: 5 },
                },
            },
            {
                kind: Lexeme.True,
                text: "true",
                literal: BrsBoolean.True,
                isReserved: true,
                location: {
                    start: { line: 1, column: 6 },
                    end: { line: 1, column: 10 },
                },
            },
            {
                kind: Lexeme.Newline,
                text: "\n",
                isReserved: false,
                location: {
                    start: { line: 1, column: 10 },
                    end: { line: 1, column: 11 },
                },
            },
            // loop body isn't significant for location tracking, so helper functions are safe
            identifier("Rnd"),
            token(Lexeme.LeftParen, "("),
            token(Lexeme.Integer, "0", new Int32(0)),
            token(Lexeme.RightParen, ")"),
            token(Lexeme.Newline, "\n"),

            {
                kind: Lexeme.EndWhile,
                text: "end while",
                isReserved: false,
                location: {
                    start: { line: 3, column: 0 },
                    end: { line: 3, column: 9 },
                },
            },
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements.length).toBe(1);
        expect(statements[0].location).toEqual({
            start: { line: 1, column: 0 },
            end: { line: 3, column: 9 },
        });
    });
});
