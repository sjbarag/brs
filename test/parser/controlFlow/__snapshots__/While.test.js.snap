// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser while statements nested 1`] = `
Array [
  While {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": 12,
          "line": 8,
        },
        "file": "",
        "start": Object {
          "column": 26,
          "line": 2,
        },
      },
      "statements": Array [
        While {
          "body": Block {
            "location": Object {
              "end": Object {
                "column": 16,
                "line": 6,
              },
              "file": "",
              "start": Object {
                "column": 30,
                "line": 3,
              },
            },
            "statements": Array [
              While {
                "body": Block {
                  "location": Object {
                    "end": Object {
                      "column": 43,
                      "line": 4,
                    },
                    "file": "",
                    "start": Object {
                      "column": 35,
                      "line": 4,
                    },
                  },
                  "statements": Array [
                    Increment {
                      "token": Object {
                        "isReserved": false,
                        "kind": "PlusPlus",
                        "literal": undefined,
                        "location": Object {
                          "end": Object {
                            "column": 40,
                            "line": 4,
                          },
                          "file": "",
                          "start": Object {
                            "column": 38,
                            "line": 4,
                          },
                        },
                        "text": "++",
                      },
                      "type": "Increment",
                      "value": Variable {
                        "name": Object {
                          "isReserved": false,
                          "kind": "Identifier",
                          "literal": undefined,
                          "location": Object {
                            "end": Object {
                              "column": 38,
                              "line": 4,
                            },
                            "file": "",
                            "start": Object {
                              "column": 37,
                              "line": 4,
                            },
                          },
                          "text": "k",
                        },
                        "type": "Variable",
                      },
                    },
                  ],
                  "type": "Block",
                },
                "condition": Binary {
                  "left": Variable {
                    "name": Object {
                      "isReserved": false,
                      "kind": "Identifier",
                      "literal": undefined,
                      "location": Object {
                        "end": Object {
                          "column": 27,
                          "line": 4,
                        },
                        "file": "",
                        "start": Object {
                          "column": 26,
                          "line": 4,
                        },
                      },
                      "text": "k",
                    },
                    "type": "Variable",
                  },
                  "right": Literal {
                    "_location": Object {
                      "end": Object {
                        "column": 34,
                        "line": 4,
                      },
                      "file": "",
                      "start": Object {
                        "column": 30,
                        "line": 4,
                      },
                    },
                    "type": "Literal",
                    "value": Int32 {
                      "kind": 4,
                      "value": 1000,
                    },
                  },
                  "token": Object {
                    "isReserved": false,
                    "kind": "Less",
                    "literal": undefined,
                    "location": Object {
                      "end": Object {
                        "column": 29,
                        "line": 4,
                      },
                      "file": "",
                      "start": Object {
                        "column": 28,
                        "line": 4,
                      },
                    },
                    "text": "<",
                  },
                  "type": "Binary",
                },
                "tokens": Object {
                  "endWhile": Object {
                    "isReserved": false,
                    "kind": "EndWhile",
                    "literal": undefined,
                    "location": Object {
                      "end": Object {
                        "column": 52,
                        "line": 4,
                      },
                      "file": "",
                      "start": Object {
                        "column": 43,
                        "line": 4,
                      },
                    },
                    "text": "end while",
                  },
                  "while": Object {
                    "isReserved": true,
                    "kind": "While",
                    "literal": undefined,
                    "location": Object {
                      "end": Object {
                        "column": 25,
                        "line": 4,
                      },
                      "file": "",
                      "start": Object {
                        "column": 20,
                        "line": 4,
                      },
                    },
                    "text": "while",
                  },
                },
                "type": "While",
              },
              Increment {
                "token": Object {
                  "isReserved": false,
                  "kind": "PlusPlus",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 23,
                      "line": 5,
                    },
                    "file": "",
                    "start": Object {
                      "column": 21,
                      "line": 5,
                    },
                  },
                  "text": "++",
                },
                "type": "Increment",
                "value": Variable {
                  "name": Object {
                    "isReserved": false,
                    "kind": "Identifier",
                    "literal": undefined,
                    "location": Object {
                      "end": Object {
                        "column": 21,
                        "line": 5,
                      },
                      "file": "",
                      "start": Object {
                        "column": 20,
                        "line": 5,
                      },
                    },
                    "text": "j",
                  },
                  "type": "Variable",
                },
              },
            ],
            "type": "Block",
          },
          "condition": Binary {
            "left": Variable {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 23,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 22,
                    "line": 3,
                  },
                },
                "text": "j",
              },
              "type": "Variable",
            },
            "right": Literal {
              "_location": Object {
                "end": Object {
                  "column": 30,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 26,
                  "line": 3,
                },
              },
              "type": "Literal",
              "value": Int32 {
                "kind": 4,
                "value": 1000,
              },
            },
            "token": Object {
              "isReserved": false,
              "kind": "Less",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 25,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 24,
                  "line": 3,
                },
              },
              "text": "<",
            },
            "type": "Binary",
          },
          "tokens": Object {
            "endWhile": Object {
              "isReserved": false,
              "kind": "EndWhile",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 25,
                  "line": 6,
                },
                "file": "",
                "start": Object {
                  "column": 16,
                  "line": 6,
                },
              },
              "text": "end while",
            },
            "while": Object {
              "isReserved": true,
              "kind": "While",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 21,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 16,
                  "line": 3,
                },
              },
              "text": "while",
            },
          },
          "type": "While",
        },
        Increment {
          "token": Object {
            "isReserved": false,
            "kind": "PlusPlus",
            "literal": undefined,
            "location": Object {
              "end": Object {
                "column": 19,
                "line": 7,
              },
              "file": "",
              "start": Object {
                "column": 17,
                "line": 7,
              },
            },
            "text": "++",
          },
          "type": "Increment",
          "value": Variable {
            "name": Object {
              "isReserved": false,
              "kind": "Identifier",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 17,
                  "line": 7,
                },
                "file": "",
                "start": Object {
                  "column": 16,
                  "line": 7,
                },
              },
              "text": "i",
            },
            "type": "Variable",
          },
        },
      ],
      "type": "Block",
    },
    "condition": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 18,
              "line": 2,
            },
          },
          "text": "i",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": 26,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 1000,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Less",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 21,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 20,
            "line": 2,
          },
        },
        "text": "<",
      },
      "type": "Binary",
    },
    "tokens": Object {
      "endWhile": Object {
        "isReserved": false,
        "kind": "EndWhile",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 21,
            "line": 8,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 8,
          },
        },
        "text": "end while",
      },
      "while": Object {
        "isReserved": true,
        "kind": "While",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 17,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "while",
      },
    },
    "type": "While",
  },
]
`;

exports[`parser while statements while with exit 1`] = `
Array [
  While {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [
        Print {
          "expressions": Array [
            Literal {
              "_location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "type": "Literal",
              "value": BrsString {
                "kind": 3,
                "value": "looping",
              },
            },
          ],
          "tokens": Object {
            "print": Object {
              "isReserved": true,
              "kind": "Print",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "print",
            },
          },
          "type": "Print",
        },
        ExitWhile {
          "tokens": Object {
            "exitWhile": Object {
              "isReserved": false,
              "kind": "ExitWhile",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "exit while",
            },
          },
          "type": "ExitWhile",
        },
      ],
      "type": "Block",
    },
    "condition": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": BrsBoolean {
        "kind": 2,
        "value": true,
      },
    },
    "tokens": Object {
      "endWhile": Object {
        "isReserved": false,
        "kind": "EndWhile",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end while",
      },
      "while": Object {
        "isReserved": true,
        "kind": "While",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "while",
      },
    },
    "type": "While",
  },
]
`;

exports[`parser while statements while without exit 1`] = `
Array [
  While {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [
        Print {
          "expressions": Array [
            Literal {
              "_location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "type": "Literal",
              "value": BrsString {
                "kind": 3,
                "value": "looping",
              },
            },
          ],
          "tokens": Object {
            "print": Object {
              "isReserved": true,
              "kind": "Print",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "print",
            },
          },
          "type": "Print",
        },
      ],
      "type": "Block",
    },
    "condition": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": BrsBoolean {
        "kind": 2,
        "value": true,
      },
    },
    "tokens": Object {
      "endWhile": Object {
        "isReserved": false,
        "kind": "EndWhile",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end while",
      },
      "while": Object {
        "isReserved": true,
        "kind": "While",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "while",
      },
    },
    "type": "While",
  },
]
`;
